### PR TITLE
Wait for Requeue condition before push to heap.

### LINKED ISSUE #2291: [WaitForPodsReady] Wrong transitions of the Requeued condition when using WaitForPodsReady
/kind bug

**What happened**:

Deactivation (setting `.status.active: false`) is delayed unnecessarily until `Requeued: true`. This is unnecessary, because the fate of the workload is already known once the timeout is exceeded and `requeuingState.count = backoffLimitCount`.

Also, when `Requeue: true` race conditions are possible with scheduler which tries to re-admit.

**What you expected to happen**:

The workload should be deactivated as soon as it exceeds PodsReady timeout and requeuingState.count = backoffLimitCount`.

**How to reproduce it (as minimally and precisely as possible)**:

1. Enable WaitForPodsReady with the following, or analogous, config:
```yaml
      waitForPodsReady:
        enable: true
        timeout: 30s
        blockAdmission: false
        requeuingStrategy:
          timestamp: Eviction
          backoffLimitCount: 3 # null indicates infinite requeuing
          backoffBaseSeconds: 10
```
2. Create a watch for the workloads `kubectl get workload -w --output-watch-events -ocustom-columns=EVENT:.type,NAME:.object.metadata.name,ACTIVE:.object.spec.active,CONDITIONS:.object.status.conditions | ts "%Y-%m-%d %H:%M:%.S"`

3. Issue: the `active: false` is only set when `Requeued; true`, which is delayed.

**Anything else we need to know?**:

We discussed slack with @tenzen-y to implement the following flow as a bugfix:
1. increment ` requeuingState.count` to `backoffLimitCount+1`   to exceed `backoffLimitCount` when handling PodsReady timeout
2. deactivate, set `active: false`
3. set `Evicted: true` with `WorkloadInactive` condition and message

Note that 2., and 3. already happen.

We also discussed, that in a longer run, for  0.8 we may additionally introduce `DeactivatedTarget` condition to store the message.

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
When backoff finished we have race condition on the Workload Reconcile and the Scheduler. The Scheduler using old version of the Workload (version of workload before updated Requeue condition to true) and rewriting conditions (already updated by reconciler) when admitting. So we need to wait before Requeue condition to be true.

#### Which issue(s) this PR fixes:

Fixes #2291 

#### Special notes for your reviewer:

```bash
> kubectl get workload -w --output-watch-events -ocustom-columns=EVENT:.type,NAME:.object.metadata.name,ACTIVE:.object.spec.active,CONDITIONS:.object.status.conditions | ts "%Y-%m-%d %H:%M:%.S"

2024-06-04 09:06:58.564580 EVENT   NAME                   ACTIVE   CONDITIONS
# 1.1. Created:
2024-06-04 09:06:58.564735 ADDED   job-sample-job-992ea   true     <none>
# 1.2. Set PodsReady=False:
2024-06-04 09:06:58.572070 MODIFIED   job-sample-job-992ea   true     [map[lastTransitionTime:2024-06-04T07:06:58Z message:Not all pods are ready or succeeded observedGeneration:1 reason:PodsReady status:False type:PodsReady]]
# 1.3. Set QuotaReserver=True and Admitted=True:
2024-06-04 09:06:58.583652 MODIFIED   job-sample-job-992ea   true     [map[lastTransitionTime:2024-06-04T07:06:58Z message:Not all pods are ready or succeeded observedGeneration:1 reason:PodsReady status:False type:PodsReady] map[lastTransitionTime:2024-06-04T07:06:58Z message:Quota reserved in ClusterQueue cluster-queue observedGeneration:1 reason:QuotaReserved status:True type:QuotaReserved] map[lastTransitionTime:2024-06-04T07:06:58Z message:The workload is admitted observedGeneration:1 reason:Admitted status:True type:Admitted]]

# 2.1. Pods not ready Timeout first time (after 1 minute). Evict=true:
# https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/core/workload_controller.go#L483-L485
2024-06-04 09:07:58.013507 MODIFIED   job-sample-job-992ea   true     [map[lastTransitionTime:2024-06-04T07:06:58Z message:Not all pods are ready or succeeded observedGeneration:1 reason:PodsReady status:False type:PodsReady] map[lastTransitionTime:2024-06-04T07:06:58Z message:Quota reserved in ClusterQueue cluster-queue observedGeneration:1 reason:QuotaReserved status:True type:QuotaReserved] map[lastTransitionTime:2024-06-04T07:07:58Z message:Exceeded the PodsReady timeout default/job-sample-job-992ea observedGeneration:1 reason:PodsReadyTimeout status:True type:Evicted] map[lastTransitionTime:2024-06-04T07:06:58Z message:The workload is admitted observedGeneration:1 reason:Admitted status:True type:Admitted]]
# 2.2. Handle eviction. Requeued=False, Admitted=False.
# https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/reconciler.go#L424-L447.
2024-06-04 09:07:59.042140 MODIFIED   job-sample-job-992ea   true     [map[lastTransitionTime:2024-06-04T07:06:58Z message:Not all pods are ready or succeeded observedGeneration:1 reason:PodsReady status:False type:PodsReady] map[lastTransitionTime:2024-06-04T07:07:59Z message:Exceeded the PodsReady timeout default/job-sample-job-992ea observedGeneration:1 reason:Pending status:False type:QuotaReserved] map[lastTransitionTime:2024-06-04T07:07:58Z message:Exceeded the PodsReady timeout default/job-sample-job-992ea observedGeneration:1 reason:PodsReadyTimeout status:True type:Evicted] map[lastTransitionTime:2024-06-04T07:07:59Z message:The workload has no reservation observedGeneration:1 reason:NoReservation status:False type:Admitted] map[lastTransitionTime:2024-06-04T07:07:59Z message:Exceeded the PodsReady timeout default/job-sample-job-992ea observedGeneration:1 reason:PodsReadyTimeout status:False type:Requeued]]

# 3.1. Readmit after backoff (after 1 minute) Requeued=True.
# https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/core/workload_controller.go#L157-L181.
2024-06-04 09:08:58.014986 MODIFIED   job-sample-job-992ea   true     [map[lastTransitionTime:2024-06-04T07:06:58Z message:Not all pods are ready or succeeded observedGeneration:1 reason:PodsReady status:False type:PodsReady] map[lastTransitionTime:2024-06-04T07:07:59Z message:Exceeded the PodsReady timeout default/job-sample-job-992ea observedGeneration:1 reason:Pending status:False type:QuotaReserved] map[lastTransitionTime:2024-06-04T07:07:58Z message:Exceeded the PodsReady timeout default/job-sample-job-992ea observedGeneration:1 reason:PodsReadyTimeout status:True type:Evicted] map[lastTransitionTime:2024-06-04T07:07:59Z message:The workload has no reservation observedGeneration:1 reason:NoReservation status:False type:Admitted] map[lastTransitionTime:2024-06-04T07:08:58Z message:The workload backoff was finished observedGeneration:1 reason:BackoffFinished status:True type:Requeued]]
# 3.2. (https://github.com/kubernetes-sigs/kueue/blob/main/pkg/scheduler/scheduler.go#L530). Requeued=True:
# BUG! (it should not set Requeued=False second time).
2024-06-04 09:08:58.027117 MODIFIED   job-sample-job-992ea   true     [map[lastTransitionTime:2024-06-04T07:06:58Z message:Not all pods are ready or succeeded observedGeneration:1 reason:PodsReady status:False type:PodsReady] map[lastTransitionTime:2024-06-04T07:08:58Z message:Quota reserved in ClusterQueue cluster-queue observedGeneration:1 reason:QuotaReserved status:True type:QuotaReserved] map[lastTransitionTime:2024-06-04T07:08:58Z message:Previously: Exceeded the PodsReady timeout default/job-sample-job-992ea observedGeneration:1 reason:QuotaReserved status:False type:Evicted] map[lastTransitionTime:2024-06-04T07:08:58Z message:The workload is admitted observedGeneration:1 reason:Admitted status:True type:Admitted] map[lastTransitionTime:2024-06-04T07:07:59Z message:Exceeded the PodsReady timeout default/job-sample-job-992ea observedGeneration:1 reason:PodsReadyTimeout status:False type:Requeued]]
# 3.3. https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/core/workload_controller.go#L157-L181). Requeued=True.
# BUG! (it should not set Requeued=True second time).
2024-06-04 09:08:58.042313 MODIFIED   job-sample-job-992ea   true     [map[lastTransitionTime:2024-06-04T07:06:58Z message:Not all pods are ready or succeeded observedGeneration:1 reason:PodsReady status:False type:PodsReady] map[lastTransitionTime:2024-06-04T07:08:58Z message:Quota reserved in ClusterQueue cluster-queue observedGeneration:1 reason:QuotaReserved status:True type:QuotaReserved] map[lastTransitionTime:2024-06-04T07:08:58Z message:Previously: Exceeded the PodsReady timeout default/job-sample-job-992ea observedGeneration:1 reason:QuotaReserved status:False type:Evicted] map[lastTransitionTime:2024-06-04T07:08:58Z message:The workload is admitted observedGeneration:1 reason:Admitted status:True type:Admitted] map[lastTransitionTime:2024-06-04T07:08:58Z message:The workload backoff was finished observedGeneration:1 reason:BackoffFinished status:True type:Requeued]]

# 4.1. Deactivation due to Timeout second time. Active=False:
https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/core/workload_controller.go#L458-L466.
2024-06-04 09:09:58.013151 MODIFIED   job-sample-job-992ea   false    [map[lastTransitionTime:2024-06-04T07:06:58Z message:Not all pods are ready or succeeded observedGeneration:1 reason:PodsReady status:False type:PodsReady] map[lastTransitionTime:2024-06-04T07:08:58Z message:Quota reserved in ClusterQueue cluster-queue observedGeneration:1 reason:QuotaReserved status:True type:QuotaReserved] map[lastTransitionTime:2024-06-04T07:08:58Z message:Previously: Exceeded the PodsReady timeout default/job-sample-job-992ea observedGeneration:1 reason:QuotaReserved status:False type:Evicted] map[lastTransitionTime:2024-06-04T07:08:58Z message:The workload is admitted observedGeneration:1 reason:Admitted status:True type:Admitted] map[lastTransitionTime:2024-06-04T07:08:58Z message:The workload backoff was finished observedGeneration:1 reason:BackoffFinished status:True type:Requeued]]
# 4.2. Handle Deactivation 
# https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/core/workload_controller.go#L183-L214:
2024-06-04 09:09:58.025217 MODIFIED   job-sample-job-992ea   false    [map[lastTransitionTime:2024-06-04T07:06:58Z message:Not all pods are ready or succeeded observedGeneration:1 reason:PodsReady status:False type:PodsReady] map[lastTransitionTime:2024-06-04T07:08:58Z message:Quota reserved in ClusterQueue cluster-queue observedGeneration:1 reason:QuotaReserved status:True type:QuotaReserved] map[lastTransitionTime:2024-06-04T07:09:58Z message:The workload is deactivated due to exceeding the maximum number of re-queuing retries observedGeneration:2 reason:InactiveWorkload status:True type:Evicted] map[lastTransitionTime:2024-06-04T07:08:58Z message:The workload is admitted observedGeneration:1 reason:Admitted status:True type:Admitted] map[lastTransitionTime:2024-06-04T07:08:58Z message:The workload backoff was finished observedGeneration:1 reason:BackoffFinished status:True type:Requeued]]
# 4.3. Set Requeued=False after eviction
# https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/reconciler.go#L424-L447
2024-06-04 09:09:59.052521 MODIFIED   job-sample-job-992ea   false    [map[lastTransitionTime:2024-06-04T07:06:58Z message:Not all pods are ready or succeeded observedGeneration:1 reason:PodsReady status:False type:PodsReady] map[lastTransitionTime:2024-06-04T07:09:59Z message:The workload is deactivated due to exceeding the maximum number of re-queuing retries observedGeneration:2 reason:Pending status:False type:QuotaReserved] map[lastTransitionTime:2024-06-04T07:09:58Z message:The workload is deactivated due to exceeding the maximum number of re-queuing retries observedGeneration:2 reason:InactiveWorkload status:True type:Evicted] map[lastTransitionTime:2024-06-04T07:09:59Z message:The workload has no reservation observedGeneration:2 reason:NoReservation status:False type:Admitted] map[lastTransitionTime:2024-06-04T07:09:59Z message:The workload is deactivated due to exceeding the maximum number of re-queuing retries observedGeneration:2 reason:InactiveWorkload status:False type:Requeued]]
```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/queue/cluster_queue_test.go [hunk 1]
```diff
@@ -86,6 +90,36 @@ func Test_PushOrUpdate(t *testing.T) {
...
 						Status: metav1.ConditionTrue,
 					}).
+					Condition(metav1.Condition{
+						Type:   kueue.WorkloadRequeued,
+						Status: metav1.ConditionFalse,
+					}).
+					Obj()),
+			},
+		},
+		"workload is still have Requeued=true": {
```
Comment (mimowo) [inline pkg/queue/cluster_queue_test.go:100]: Is this description accurate, the workload has `Requeued: false`. Please try to capture the intention of the new test case in the name.

**Review by mimowo — COMMENTED** [review]
/approve
LGTM, just a nit about the unit tests.
@tenzen-y anything to add?

File: pkg/queue/cluster_queue_test.go (hunk 1 above)
Comment (mbobrovskyi) [reply pkg/queue/cluster_queue_test.go]: Thanks. Fixed.

**Comment by tenzen-y** [comment]
> /approve LGTM, just a nit about the unit tests. @tenzen-y anything to add?

ACK

File: pkg/queue/cluster_queue_test.go [hunk 2]
```diff
@@ -86,6 +90,36 @@ func Test_PushOrUpdate(t *testing.T) {
...
+						Reason: kueue.WorkloadEvictedByPodsReadyTimeout,
+						Status: metav1.ConditionTrue,
+					}).
+					Condition(metav1.Condition{
+						Type:   kueue.WorkloadRequeued,
+						Status: metav1.ConditionFalse,
+					}).
 					Obj()),
 			},
 		},
```
Comment (tenzen-y) [inline pkg/queue/cluster_queue_test.go:125]: Could you add a case to verify the situation of changed Requeued Condition?

**Review by tenzen-y — COMMENTED** [review]
Thank you for investigating this problem.
Basically, lgtm

/cherry-pick release-0.7

File: pkg/queue/cluster_queue_test.go (hunk 2 above)
Comment (mbobrovskyi) [reply pkg/queue/cluster_queue_test.go:125]: Done

**Review by tenzen-y — COMMENTED** [review]
Thank you
/lgtm
/approve

**Pull request merged**